File: sdk/python/src/plinth/tokens.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 The Plinth Authors
"""Offline token counting and cost estimation helpers.

Plinth intentionally avoids round-tripping to a model just to count
tokens. We use ``tiktoken`` with the ``cl100k_base`` encoding, which is
a close-enough approximation for Anthropic's BPE tokenizer for the
budgeting and observability use cases the SDK targets.

The encoding is loaded lazily and cached at module level so repeated
``count`` calls do not pay the constructor cost.
"""

from __future__ import annotations

from functools import lru_cache
from typing import TYPE_CHECKING

if TYPE_CHECKING:  # pragma: no cover - import only for type checkers
    import tiktoken


# ---------------------------------------------------------------------------
# Pricing constants — keep these obvious so they're easy to refresh.
# ---------------------------------------------------------------------------

#: Anthropic Claude Sonnet input pricing in USD per 1M tokens.
SONNET_INPUT_USD_PER_MTOK: float = 3.0

#: Anthropic Claude Sonnet output pricing in USD per 1M tokens.
SONNET_OUTPUT_USD_PER_MTOK: float = 15.0

#: The tiktoken encoding name we standardise on. ``cl100k_base`` is the
#: encoding used by recent OpenAI models and is the closest publicly
#: available BPE to Anthropic's tokenizer.
ENCODING_NAME: str = "cl100k_base"


@lru_cache(maxsize=1)
def _get_encoding() -> tiktoken.Encoding:
    """Load and cache the tiktoken encoding.

    We use ``functools.lru_cache`` rather than a module-level global so
    that test code can clear the cache in fixtures without monkey-
    patching internals.
    """
    import tiktoken  # local import keeps SDK import time low

    return tiktoken.get_encoding(ENCODING_NAME)


def count(text: str) -> int:
    """Return the number of tokens in ``text``.

    Args:
        text: Any string. Empty strings return ``0``.

    Returns:
        The number of tokens produced by the ``cl100k_base`` encoding.
    """
    if not text:
        return 0
    encoding = _get_encoding()
    return len(encoding.encode(text))


def estimate_cost(prompt_tokens: int, completion_tokens: int = 0) -> float:
    """Estimate the USD cost of a Sonnet request.

    Args:
        prompt_tokens: Number of input/prompt tokens.
        completion_tokens: Number of generated tokens. Defaults to 0 so
            callers can estimate prompt-only cost.

    Returns:
        The estimated cost in USD, computed at Sonnet pricing.
    """
    if prompt_tokens < 0 or completion_tokens < 0:
        raise ValueError("Token counts must be non-negative")
    input_cost = prompt_tokens * SONNET_INPUT_USD_PER_MTOK / 1_000_000
    output_cost = completion_tokens * SONNET_OUTPUT_USD_PER_MTOK / 1_000_000
    return input_cost + output_cost


__all__ = [
    "ENCODING_NAME",
    "SONNET_INPUT_USD_PER_MTOK",
    "SONNET_OUTPUT_USD_PER_MTOK",
    "count",
    "estimate_cost",
]

File: sdk/go/plinth/tokens.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The Plinth Authors

package plinth

import (
	"strings"
	"unicode"
)

// Token-counting and cost-estimation helpers. Mirrors
// sdk/python/src/plinth/tokens.py and sdk/typescript/src/tokens.ts.
//
// v0.1 of the Go SDK uses a deliberately lightweight approximation:
// full tiktoken / gpt-tokenizer would pull in a multi-MB BPE table and
// mean a non-stdlib dependency. The approximation:
//
//   1. Splits text into "words" (runs of letters/digits) and punctuation.
//   2. For each word, estimates ceil(len(word) / 4) tokens — matches
//      cl100k's average byte-pair encoding density on English / code.
//   3. Counts every punctuation rune as a single token.
//   4. Treats whitespace runs (newlines, multiple spaces) as one token
//      each — newlines are often their own BPE token in cl100k.
//
// The result is within ~5–10% of cl100k_base on typical English /
// code prompts. For exact accounting (billing, hard token caps),
// callers should reach for a full BPE implementation.

// SonnetInputUSDPerMTok is the Sonnet 4.x input price per million
// tokens. Mirrors the Python SDK's constants.
const SonnetInputUSDPerMTok = 3.0

// SonnetOutputUSDPerMTok is the Sonnet 4.x output price per million
// tokens.
const SonnetOutputUSDPerMTok = 15.0

// CountTokens returns an approximate cl100k token count for text.
//
// See the package doc comment above for the algorithm and accuracy
// caveats. Returns 0 for the empty string.
func CountTokens(text string) int {
	if text == "" {
		return 0
	}
	tokens := 0
	var word strings.Builder

	flushWord := func() {
		if word.Len() == 0 {
			return
		}
		// Average BPE token length on cl100k_base is ~4 chars for
		// English; cap at 1 for very short words.
		n := (word.Len() + 3) / 4
		if n < 1 {
			n = 1
		}
		tokens += n
		word.Reset()
	}

	inWhitespace := false
	for _, r := range text {
		switch {
		case unicode.IsSpace(r):
			flushWord()
			if !inWhitespace {
				// One token per whitespace run.
				tokens++
				inWhitespace = true
			}
			// Newlines are usually their own BPE token in cl100k —
			// add an extra so multi-line prompts don't undercount.
			if r == '\n' {
				tokens++
			}
		case unicode.IsLetter(r) || unicode.IsDigit(r):
			word.WriteRune(r)
			inWhitespace = false
		default:
			// Punctuation, symbols, control chars: each gets one token.
			flushWord()
			tokens++
			inWhitespace = false
		}
	}
	flushWord()
	if tokens == 0 {
		// Defensive — non-empty input should always produce ≥1 token.
		return 1
	}
	return tokens
}

// EstimateCost returns a USD estimate for a Sonnet-class request given
// its input and output token counts. Pass 0 for outputTokens to get
// the prompt-only cost.
func EstimateCost(inputTokens, outputTokens int) float64 {
	in := float64(inputTokens) * SonnetInputUSDPerMTok / 1_000_000.0
	out := float64(outputTokens) * SonnetOutputUSDPerMTok / 1_000_000.0
	return in + out
}
